blacksheep/server/websocket.py

```diff
@@ -0,0 +1,148 @@
+import json
+
+from enum import Enum
+from functools import wraps
+from typing import (
+    Any,
+    AnyStr,
+    Callable,
+    List,
+    MutableMapping,
+    Optional,
+)
+
+
+class WebSocketDisconnect(Exception):
+    def __init__(self, code: int = 1000):
+        self.code = code
+
+
+class WebSocketState(Enum):
+    CONNECTING = 0
+    CONNECTED = 1
+    DISCONNECTED = 2
+
+
+class MessageMode(str, Enum):
+    TEXT = "text"
+    BYTES = "bytes"
+
+
+class WebSocket:
+    def __init__(
+            self,
+            scope: MutableMapping[str, Any],
+            receive: Callable,
+            send: Callable
+    ):
+        assert scope["type"] == "websocket"
+
+        self._scope = scope
+        self._receive = self._wrap_receive(receive)
+        self._send = send
+        self.route_values = {}
+
+        self.client_state = WebSocketState.CONNECTING
+        self.application_state = WebSocketState.CONNECTING
+
+    async def connect(self) -> None:
+        message = await self._receive()
+        assert message["type"] == "websocket.connect"
+
+        self.client_state = WebSocketState.CONNECTED
+
+    async def accept(
+            self,
+            headers: Optional[List] = None,
+            subprotocol: str = None
+    ) -> None:
+        assert self.client_state == WebSocketState.CONNECTING
+        await self.connect()
+
+        headers = headers or []
+        self.application_state = WebSocketState.CONNECTED
+
+        message = {
+            "type": "websocket.accept",
+            "headers": headers,
+            "subprotocol": subprotocol
+        }
+
+        await self._send(message)
+
+    async def receive(self) -> MutableMapping[str, AnyStr]:
+        assert self.application_state == WebSocketState.CONNECTED
```

Comment:
I agree with that, but I'm not sure it behaves like that right now. 

Take a look at the `__init__` method of the `WebSocket` class. We wrap a `receive` function with the `_wrap_receive` method. Every received message is checked for a disconnect event before returning it to the caller. In event of disconnect, the application and the client state are set to DISCONNECTED and the exception is raised. 

Thus, if the client disconnects, we already raise a `WebSocketDisconnect` error. So this assertion could fail only if a user of the class forgot to call the `accept` method before awaiting messages. That seems to me like an appropriate use of the assert statement.

Sorry, if I'm missing something. Can you confirm that this assertion could raise an exception if the client disconnects?

Reply:
I also noticed that sometimes `websockets` raises an exception when the client disconnects with 1001 (like just closing the browser tab). There's an open issue about that: https://github.com/aaugustin/websockets/issues/1117 So I guess it's not our fault. 🤔

Reply:
Maybe I was doing something wrong myself, because I was getting assertion errors on the receive method. On my local environment, while trying your demo, I modified it a little to get messages from one client delivered to other clients. I will share the details with you as soon as I can. Anyway raising `TypeError` or a custom exception won't hurt. 😄

Reply:
Do you feel like replacing only this one assert statement with a proper exception or every assert in the module? Maybe, only these asserts, that check for a correct connection state?

Reply:
I was thinking only the asserts that check the connection state (I don't recall how many they are, two or three?). I was also thinking maybe for exceptions:

```python
class WebSocketError(Exception):
    """Base class for all web sockets errors"""


class InvalidWebSocketStateError(WebSocketError):

    def __init__(self, current_state: WebSocketState, expected_state: WebSocketState):
        super().__init__(f"The current state of the WebSocket is invalid. Expected state: {expected_state}")
        self.current_state = current_state
        self.expected_state = expected_state


class WebSocketDisconnectedError(InvalidWebSocketStateError):

    def __init__(self, current_state: WebSocketState):
        super().__init__(current_state, WebSocketState.Connected)
```